### PR TITLE
[type] Add hana::typeid_ and deprecate hana::decltype_

### PR DESCRIPTION
**Pull request by ldionne**
Candidate for satisfying #290. I'm still unsure whether using `typeof` would be better, though.

### CODE REVIEW

File: include/boost/hana/fwd/type.hpp [hunk 1]
```diff
@@ -121,6 +121,11 @@ BOOST_HANA_NAMESPACE_BEGIN
     //! `decltype` keyword, lifted to Hana.
     //! @relates hana::type
     //!
+    //! @deprecated
+    //! The semantics of `decltype_` are can be confusing, and `hana::typeid_`
```
Comment (viboes) [inline include/boost/hana/fwd/type.hpp]: s/are can/can/

Comment (ldionne) [inline include/boost/hana/fwd/type.hpp]: Thanks!

**Comment by ricejasonf** [comment]
`typeof` :+1:

**Comment by ldionne** [comment]
I decided against `typeof`, even though I prefer it, because it clashes with a GCC extension of the same name. When compiling with `std=gnu++14`, using `typeof` anywhere in your program causes a compilation error.

**Pull request merged**